client/serviceregistration/wrapper/wrapper.go

```diff
@@ -0,0 +1,131 @@
+package wrapper
+
+import (
+	"fmt"
+
+	"github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/nomad/client/serviceregistration"
+	"github.com/hashicorp/nomad/nomad/structs"
+)
+
+// HandlerWrapper is used to wrap service registration implementations of the
+// Handler interface. We do not use a map or similar to store the handlers, so
+// we can avoid having to use a lock. This may need to be updated if we ever
+// support additional registration providers.
+type HandlerWrapper struct {
+	log hclog.Logger
+
+	// consulServiceProvider is the handler for services where Consul is the
+	// provider. This provider is always created and available.
+	consulServiceProvider serviceregistration.Handler
+
+	// nomadServiceProvider is the handler for services where Nomad is the
+	// provider.
+	nomadServiceProvider serviceregistration.Handler
+}
+
+// NewHandlerWrapper configures and returns a HandlerWrapper for use within
+// client hooks that need to interact with service and check registrations. It
+// mimics the serviceregistration.Handler interface, but returns the
+// implementation to allow future flexibility and is initially only intended
+// for use with the alloc and task runner service hooks.
+func NewHandlerWrapper(
+	log hclog.Logger, consulProvider, nomadProvider serviceregistration.Handler) *HandlerWrapper {
+	return &HandlerWrapper{
+		log:                   log,
+		nomadServiceProvider:  nomadProvider,
+		consulServiceProvider: consulProvider,
+	}
+}
+
+// RegisterWorkload wraps the serviceregistration.Handler RegisterWorkload
+// function. It determines which backend provider to call and passes the
+// workload unless the provider is unknown, in which case an error will be
+// returned.
+func (h *HandlerWrapper) RegisterWorkload(workload *serviceregistration.WorkloadServices) error {
+
+	// Don't rely on callers to check there are no services to register.
+	if len(workload.Services) == 0 {
+		return nil
+	}
+
+	provider := workload.RegistrationProvider()
+
+	switch provider {
+	case structs.ServiceProviderNomad:
+		return h.nomadServiceProvider.RegisterWorkload(workload)
+	case structs.ServiceProviderConsul:
+		return h.consulServiceProvider.RegisterWorkload(workload)
```

Comment:
Could we support multiple service reg providers per alloc by making WorkloadServices.Handlers a `map[string][]*Service` instead of just a flat `[]*Service` slice? The mapping would be `provider -> Services.

I'm not sure we would even need this wrapper... we'd just iterate over all providers and pass them the workloads. They pluck out the services they care about, if any. I think the Update case would just work as well.

This all looks perfectly workable, so this isn't a blocker even if the idea works in theory.

Reply:
Yes potentially.

Assuming i'll follow up after the initial release with work to allow multiple providers per allocation I will keep this idea in mind. It might still be useful to have the wrapper to avoid duplicated logic within each service hook; the unknowns are why I wanted to leave this improvement till a later date so we can understand what will work best and how.